File: backend/src/protocol/packetSerialization.test.ts

/**
 * packetValidation.test.ts — Phase 10 unit tests
 *
 * Backend structural-validation-only tests. Covers valid packets, malformed
 * JSON, missing required fields, invalid packet types, and confirms every
 * stage beyond structural validation is reported "deferred" — and, for the
 * backend, permanently out of scope — never "passed".
 */

import { randomUUID } from "crypto";
import { serializePacket, deserializePacket, PacketSerializationError } from "./packetSerialization.js";
import {
  validatePacketStructure,
  PACKET_VALIDATION_STAGE_ORDER,
} from "./packetvalidation.js";
import type { Packet } from "./packets.js";

function samplePacket(overrides: Partial<{
  protocolVersion: number;
  sessionId: string;
  senderId: string;
  packetId: string;
}> = {}): Packet {
  return {
    header: {
      protocolVersion: overrides.protocolVersion ?? 1,
      packetType: "HEARTBEAT",
      sessionId: overrides.sessionId ?? randomUUID(),
      senderId: overrides.senderId ?? "device-abc",
      packetId: overrides.packetId ?? randomUUID(),
      sequenceNumber: 1,
      timestamp: 1_700_000_000_000,
    },
    metadata: {
      retryCount: 0,
      priority: "NORMAL",
      messageCategory: "HEARTBEAT",
    },
    encryptedPayload: new Uint8Array([1, 2, 3]),
  };
}

describe("packetSerialization", () => {
  test("valid packet round-trips", () => {
    const packet = samplePacket();
    const json = serializePacket(packet);
    const decoded = deserializePacket(json);
    expect(decoded.header.packetId).toBe(packet.header.packetId);
    expect(decoded.header.senderId).toBe(packet.header.senderId);
    expect(Array.from(decoded.encryptedPayload)).toEqual(Array.from(packet.encryptedPayload));
  });

  test("malformed JSON throws", () => {
    expect(() => deserializePacket("{ this is not valid json")).toThrow(PacketSerializationError);
  });

  test("missing header field throws", () => {
    const json = JSON.stringify({
      metadata: { retryCount: 0, priority: "NORMAL", messageCategory: "HEARTBEAT" },
      encryptedPayload: "AQID",
    });
    expect(() => deserializePacket(json)).toThrow(PacketSerializationError);
  });

  test("missing encryptedPayload field throws", () => {
    const json = JSON.stringify({
      header: {
        protocolVersion: 1, packetType: "HEARTBEAT", sessionId: randomUUID(),
        senderId: "d1", packetId: randomUUID(), sequenceNumber: 1, timestamp: 1,
      },
      metadata: { retryCount: 0, priority: "NORMAL", messageCategory: "HEARTBEAT" },
    });
    expect(() => deserializePacket(json)).toThrow(PacketSerializationError);
  });

  test("invalid packet type throws", () => {
    const json = JSON.stringify({
      header: {
        protocolVersion: 1, packetType: "NOT_A_REAL_TYPE", sessionId: randomUUID(),
        senderId: "d1", packetId: randomUUID(), sequenceNumber: 1, timestamp: 1,
      },
      metadata: { retryCount: 0, priority: "NORMAL", messageCategory: "HEARTBEAT" },
      encryptedPayload: "AQID",
    });
    expect(() => deserializePacket(json)).toThrow(PacketSerializationError);
  });
});

describe("packetValidation (backend, structural only)", () => {
  test("stage order matches documentation exactly", () => {
    expect(PACKET_VALIDATION_STAGE_ORDER).toEqual([
      "PROTOCOL_VERSION",
      "PACKET_STRUCTURE",
      "SENDER_IDENTITY",
      "AUTHENTICATION_STATE",
      "SIGNATURE_VERIFICATION",
      "INTEGRITY_VERIFICATION",
      "FRESHNESS_VALIDATION",
      "DUPLICATE_DETECTION",
      "PAYLOAD_DECRYPTION",
      "COMMAND_EXECUTION",
    ]);
  });

  test("unsupported protocol version fails stage 1", () => {
    const packet = samplePacket({ protocolVersion: 999 });
    const report = validatePacketStructure(packet);
    expect(report.results.PROTOCOL_VERSION.status).toBe("failed");
    expect(report.passedBackendPermittedStages).toBe(false);
  });

  test("malformed packetId fails stage 2", () => {
    const packet = samplePacket({ packetId: "not-a-uuid" });
    const report = validatePacketStructure(packet);
    expect(report.results.PACKET_STRUCTURE.status).toBe("failed");
  });

  test("empty sessionId fails stage 2 as missing required field", () => {
    const packet = samplePacket({ sessionId: "" });
    const report = validatePacketStructure(packet);
    const result = report.results.PACKET_STRUCTURE;
    expect(result.status).toBe("failed");
    if (result.status === "failed" && result.reason.kind === "missingRequiredField") {
      expect(result.reason.fieldName).toBe("header.sessionId");
    } else {
      throw new Error("Expected missingRequiredField failure");
    }
  });

  test("valid packet passes backend-permitted stages", () => {
    const packet = samplePacket();
    const report = validatePacketStructure(packet);
    expect(report.passedBackendPermittedStages).toBe(true);
  });

  test("stages 3-10 are always deferred, never passed, for the backend", () => {
    const packet = samplePacket();
    const report = validatePacketStructure(packet);
    const outOfScopeStages = [
      "SENDER_IDENTITY", "AUTHENTICATION_STATE", "SIGNATURE_VERIFICATION",
      "INTEGRITY_VERIFICATION", "FRESHNESS_VALIDATION", "DUPLICATE_DETECTION",
      "PAYLOAD_DECRYPTION", "COMMAND_EXECUTION",
    ] as const;
    for (const stage of outOfScopeStages) {
      expect(report.results[stage].status).toBe("deferred");
    }
  });
});

File: android/app/src/main/java/com/macecho/app/protocol/PacketSerializationTest.kt

package com.macecho.app.protocol

import org.junit.Assert.*
import org.junit.Test
import java.util.UUID

/**
 * PacketSerializationTest.kt — Phase 10 unit tests
 *
 * Covers: valid packets, malformed packets, missing required fields,
 * invalid packet types, and validation stages 1-3. Stages 4-10 are
 * asserted to always be Deferred, never Passed.
 */
class PacketSerializationTest {

    private fun samplePacket(
        protocolVersion: Int = 1,
        sessionId: String = UUID.randomUUID().toString(),
        senderId: String = "device-abc",
        packetId: String = UUID.randomUUID().toString(),
    ): Packet = Packet(
        header = PacketHeader(
            protocolVersion = protocolVersion,
            packetType = PacketType.HEARTBEAT,
            sessionId = sessionId,
            senderId = senderId,
            packetId = packetId,
            sequenceNumber = 1L,
            timestamp = 1_700_000_000_000L,
        ),
        metadata = PacketMetadata(
            retryCount = 0,
            priority = PacketPriority.NORMAL,
            messageCategory = PacketType.HEARTBEAT,
        ),
        encryptedPayload = byteArrayOf(1, 2, 3),
    )

    // -- Serialization round-trip --

    @Test
    fun validPacketRoundTrips() {
        val packet = samplePacket()
        val json = PacketSerializer.serialize(packet)
        val decoded = PacketSerializer.deserialize(json)
        assertEquals(packet.header.packetId, decoded.header.packetId)
        assertEquals(packet.header.senderId, decoded.header.senderId)
        assertArrayEquals(packet.encryptedPayload, decoded.encryptedPayload)
    }

    // -- Malformed JSON --

    @Test(expected = PacketSerializationException.DecodingFailed::class)
    fun malformedJsonThrows() {
        PacketSerializer.deserialize("{ this is not valid json")
    }

    // -- Missing required fields --

    @Test(expected = PacketSerializationException.DecodingFailed::class)
    fun missingHeaderFieldThrows() {
        val json = """
            { "metadata": { "retryCount": 0, "priority": "NORMAL", "messageCategory": "HEARTBEAT" },
              "encryptedPayload": "AQID" }
        """.trimIndent()
        PacketSerializer.deserialize(json)
    }

    @Test(expected = PacketSerializationException.DecodingFailed::class)
    fun missingPayloadFieldThrows() {
        val json = """
            { "header": { "protocolVersion": 1, "packetType": "HEARTBEAT",
                "sessionId": "${UUID.randomUUID()}", "senderId": "d1",
                "packetId": "${UUID.randomUUID()}", "sequenceNumber": 1, "timestamp": 1 },
              "metadata": { "retryCount": 0, "priority": "NORMAL", "messageCategory": "HEARTBEAT" } }
        """.trimIndent()
        PacketSerializer.deserialize(json)
    }

    // -- Invalid packet type --

    @Test(expected = PacketSerializationException.DecodingFailed::class)
    fun invalidPacketTypeThrows() {
        val json = """
            { "header": { "protocolVersion": 1, "packetType": "NOT_A_REAL_TYPE",
                "sessionId": "${UUID.randomUUID()}", "senderId": "d1",
                "packetId": "${UUID.randomUUID()}", "sequenceNumber": 1, "timestamp": 1 },
              "metadata": { "retryCount": 0, "priority": "NORMAL", "messageCategory": "HEARTBEAT" },
              "encryptedPayload": "AQID" }
        """.trimIndent()
        PacketSerializer.deserialize(json)
    }

    // -- PacketValidator: stage 1 (protocol version) --

    @Test
    fun unsupportedProtocolVersionFails() {
        val packet = samplePacket(protocolVersion = 999)
        val report = PacketValidator.validate(packet, setOf("device-abc"))
        val result = report.results[PacketValidationStage.PROTOCOL_VERSION]
        assertTrue(result is PacketValidationStageResult.Failed)
        assertFalse(report.passedImplementedStages)
    }

    // -- PacketValidator: stage 2 (structure) --

    @Test
    fun malformedPacketIdFailsStructureStage() {
        val packet = samplePacket(packetId = "not-a-uuid")
        val report = PacketValidator.validate(packet, setOf("device-abc"))
        assertTrue(report.results[PacketValidationStage.PACKET_STRUCTURE] is PacketValidationStageResult.Failed)
    }

    @Test
    fun emptySessionIdFailsStructureStage() {
        val packet = samplePacket(sessionId = "")
        val report = PacketValidator.validate(packet, setOf("device-abc"))
        val result = report.results[PacketValidationStage.PACKET_STRUCTURE]
        assertTrue(result is PacketValidationStageResult.Failed)
        assertEquals(
            PacketValidationFailureReason.MissingRequiredField("header.sessionId"),
            (result as PacketValidationStageResult.Failed).reason,
        )
    }

    // -- PacketValidator: stage 3 (sender identity) --

    @Test
    fun unknownSenderFailsSenderIdentityStage() {
        val packet = samplePacket(senderId = "unknown-device")
        val report = PacketValidator.validate(packet, setOf("device-abc"))
        val result = report.results[PacketValidationStage.SENDER_IDENTITY]
        assertEquals(
            PacketValidationFailureReason.UnknownSenderIdentity,
            (result as PacketValidationStageResult.Failed).reason,
        )
    }

    @Test
    fun knownSenderPassesSenderIdentityStage() {
        val packet = samplePacket(senderId = "device-abc")
        val report = PacketValidator.validate(packet, setOf("device-abc"))
        assertEquals(PacketValidationStageResult.Passed, report.results[PacketValidationStage.SENDER_IDENTITY])
    }

    // -- Valid packet passes all implemented stages --

    @Test
    fun fullyValidPacketPassesImplementedStages() {
        val packet = samplePacket()
        val report = PacketValidator.validate(packet, setOf("device-abc"))
        assertTrue(report.passedImplementedStages)
        assertNull(report.firstFailure)
    }

    // -- Deferred stages are never reported as passed --

    @Test
    fun stages4Through10AreAlwaysDeferredNeverPassed() {
        val packet = samplePacket()
        val report = PacketValidator.validate(packet, setOf("device-abc"))
        val deferredStages = listOf(
            PacketValidationStage.AUTHENTICATION_STATE,
            PacketValidationStage.SIGNATURE_VERIFICATION,
            PacketValidationStage.INTEGRITY_VERIFICATION,
            PacketValidationStage.FRESHNESS_VALIDATION,
            PacketValidationStage.DUPLICATE_DETECTION,
            PacketValidationStage.PAYLOAD_DECRYPTION,
            PacketValidationStage.COMMAND_EXECUTION,
        )
        for (stage in deferredStages) {
            val result = report.results[stage]
            assertTrue("Stage $stage must be Deferred, was $result", result is PacketValidationStageResult.Deferred)
        }
    }
}